### PR TITLE
Add a Chrome badge blacklist

### PR DESCRIPTION
**Pull request by seanh**
Add a blacklist of URLs that the Chrome extension's badge will never show a number for.

### CODE REVIEW

**Comment by seanh** [comment]
<a href="https:&#x2F;&#x2F;trello.com&#x2F;c&#x2F;xgknj2RP&#x2F;109-don-t-show-chrome-badge-numbers-on-certain-pages"> Don&#39;t show Chrome badge numbers on certain pages</a>

**Comment by nickstenning** [comment]
Is there a reason why this blacklist would be distinct from the blacklist that already exists to prevent hypothesis from activating on certain sites?

**Comment by nickstenning** [comment]
Worth noting that blacklist or no blacklist, I'd be very happy to see a PR that adds a badge-specific endpoint, because it would make it a lot easier to then cache that endpoint separately from our search API.

**Comment by seanh** [comment]
I think so, yes. They're two different things: h is completely borked on this site, disable it and show the user an error/apology vs just don't show the annotation count on this page.

If a page is completely blocked though, that ought to imply that it's automatically badge-blocked as well.

They might be able to share some implementation mechanism. iirc the existing blocklist is implemented as a config file setting, built into the Chrome extension at build time. With the badge blocklist the Chrome extension is sending a request to the server to get the badge number so all we need to do is make the server respond with 0 for blocked sites, the list of sites can be server-side and editable by admins. So we could maybe put the blocked or not flag in this server response as well.

**Comment by seanh** [comment]
- [ ] Move blacklist into database
- [ ] Add admin page for blacklist
- [ ] Add server-side tests
- [ ] Support regexes in blacklist

File: h/badge/models.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+    @classmethod
+    def all(cls):
+        return cls.query.all()
+
+    @classmethod
+    def is_blocked(cls, uri):
+        """Return True if the given URI is blocked."""
+        for pattern in cls.all():
+            if fnmatch.fnmatch(uri, unicode(pattern)):
```
Comment (seanh) [inline h/badge/models.py]: This might be a lot of work for the server? If we have a lot of patterns. Let's see I guess

File: h/badge/views.py [hunk 2]
```diff
@@ -0,0 +1,33 @@
+from pyramid import httpexceptions
+
+from h import models
+from h.api.views import api_config
+from h.api.resources import Root
+from h.api import search as search_lib
```
Comment (seanh) [inline h/badge/views.py]: Hmmmm

**Comment by seanh** [comment]
@nickstenning Can be reviewed

**Comment by seanh** [comment]
@nickstenning Moved it to `LIKE` as requested

**Comment by seanh** [comment]
@nickstenning Done the table rename

File: h/badge/views.py (hunk 2 above)
Comment (nickstenning) [inline h/badge/views.py]: Ah. Well here's the thing. I don't think this should live under `/api`. It's not part of our public API. In keeping with where stuff like the session JSON blob lives, I think it should live at `/app/badge` or whatever.

I'm not going to block merging on this but we should change this before we push it out.

Comment (seanh) [inline h/badge/views.py]: Fine, I'll make this change now as a separate pr

File: h/models.py [hunk 3]
```diff
@@ -28,6 +29,7 @@
 NipsaUser = nipsa_models.NipsaUser
 Subscriptions = notification_models.Subscriptions
 User = accounts_models.User
+Blocklist = badge_models.Blocklist
```
Comment (nickstenning) [inline h/models.py]: Alphabetisation!

Comment (nickstenning) [inline h/models.py]: And you need to add it to `__all__`.

File: h/admin.py [hunk 4]
```diff
@@ -185,6 +185,38 @@ def users_index(request):
...
+    try:
+        request.db.add(models.Blocklist(uri=request.params['add']))
+    except ValueError as err:
+        request.session.flash(err.message, 'error')
+    return badge_index(request)
+
+
+@view.view_config(route_name='admin_badge',
+                  request_method='POST',
+                  request_param='remove',
```
Comment (nickstenning) [inline h/admin.py]: This is inconsistent with all the other remove views in this file. I'm sorry if this is nitpicking but inconsistency is just avoidable overhead when trying to understand the code.

Comment (seanh) [inline h/admin.py]: Alright, but the existing ones are wrong, so I've added a commit to change them as well

File: h/models.py (hunk 3 above)
Comment (seanh) [inline h/models.py]: Done

**Comment by seanh** [comment]
Rebased

**Comment by nickstenning** [comment]
Looks good to me.

**Pull request merged**